src/main/java/de/cas/mse/exercise/csv/CsvReader.java

```diff
@@ -12,37 +12,48 @@
 public class CsvReader implements Filterable {
 
     private CsvUi csvUi;
-    private List<List<String>> data;
+    private List<String> headers;
+    private List<List<String>> rows;
 
     public void run(final File csvFile) throws Exception {
         List<String> allLines = Files.readAllLines(csvFile.toPath());
         if (allLines.isEmpty()) {
             return;
         }
 
-        addHeadersToUi(parseLine(allLines.get(0)));
-        data = allLines.stream().skip(1).map(this::parseLine).collect(Collectors.toList());
-    }
+        headers = parseLine(allLines.get(0));
+        rows = getRows(allLines);
 
-    private List<String> parseLine(String line) {
-        return Arrays.asList(line.split(","));
-    }
-
-    private void addHeadersToUi(List<String> headers) {
-        headers.forEach(csvUi::addColumn);
+        render();
     }
 
     public void setCsvUi(final CsvUi csvUi) {
         this.csvUi = csvUi;
     }
 
     @Override
-    public List<List<String>> filter(int col, String word) {
-        return data.stream().filter(row -> filterApplies(row, col, word)).collect(Collectors.toList());
+    public List<List<String>> filter(int columnIndexForFiltering, String wordToContain) {
+        String filterRegex = String.format(".*\\b%s\\b.*", wordToContain);
+        return rows.stream()
+            .filter(row -> filterApplies(row, columnIndexForFiltering, filterRegex))
+            .collect(Collectors.toList());
+    }
+
+    private List<List<String>> getRows(List<String> allLines) {
+        // First line contains the header
```

Comment:
Could it be that the first line is already skipped from the input? This seems unrelated to the getRows method

Reply:
We did not want to make an additional copy of the list, nor passing Stream<> around.